pkg/controllers/user/pipeline/controller/pipelineexecution/pipelineexecution.go

```diff
@@ -384,6 +406,75 @@ func (l *Lifecycle) Remove(obj *v3.PipelineExecution) (runtime.Object, error) {
 	return obj, nil
 }
 
+func (l *Lifecycle) shouldNotify(obj *v3.PipelineExecution) (bool, error) {
+	if obj.Spec.PipelineConfig.Notification != nil {
+		if len(obj.Spec.PipelineConfig.Notification.Recipients) <= 0 {
+			return false, nil
+		}
+		condition := obj.Spec.PipelineConfig.Notification.Condition
+		if slice.ContainsString(condition, obj.Status.ExecutionState) {
+			return true, nil
+		} else if slice.ContainsString(condition, utils.ConditionChanged) {
+			_, pipelineName := ref.Parse(obj.Spec.PipelineName)
+			lastExecutionName := fmt.Sprintf("%s-%d", pipelineName, obj.Spec.Run-1)
+			lastExecution, err := l.pipelineExecutionLister.Get(obj.Namespace, lastExecutionName)
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			} else if err != nil {
+				return false, err
+			}
+			if utils.IsFinishState(lastExecution.Status.ExecutionState) &&
+				lastExecution.Status.ExecutionState != obj.Status.ExecutionState {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (l *Lifecycle) doNotify(obj *v3.PipelineExecution) (runtime.Object, error) {
+	clusterName, _ := ref.Parse(obj.Spec.ProjectName)
+	existingNotifiers, err := l.notifierLister.List(clusterName, labels.NewSelector())
+	if err != nil {
+		return obj, err
+	}
+	var toSendRecipients []struct {
+		Notifier  *mv3.Notifier
+		Recipient string
+	}
+	for _, recipient := range obj.Spec.PipelineConfig.Notification.Recipients {
+		notifierName := recipient.Notifier
+
+		for _, notifier := range existingNotifiers {
+			_, name := ref.Parse(notifierName)
+			if name == notifier.Spec.DisplayName || name == notifier.Name {
+				toSendRecipients = append(toSendRecipients, struct {
+					Notifier  *mv3.Notifier
+					Recipient string
+				}{Notifier: notifier, Recipient: recipient.Recipient})
+			}
+		}
+	}
+	message := defaultNotificationMessage(obj)
+	if obj.Spec.PipelineConfig.Notification.Message != "" {
+		message = obj.Spec.PipelineConfig.Notification.Message
+	}
+	for _, toSendRecipient := range toSendRecipients {
+		notifierMessage := &notifiers.Message{
+			Content: message,
+		}
+		if toSendRecipient.Notifier.Spec.SMTPConfig != nil {
+			repoName := getRepoNameFromURL(obj.Spec.RepositoryURL)
```

Comment:
we should check for repoName being "" as it looks like getRepoNameFromURL can return an empty string in case of a failure.

Reply:
now we don't return "" when it cannot be parsed in user/repo fmt